### PR TITLE
feat: Image Generation Module (Issues #497-501)

### PR DESCRIPTION
**Pull request by mandeepsingh2007**
##Issue: #502  

## Summary
Complete implementation of the Image Generation module for llm4s, enabling text-to-image generation and image editing across multiple cloud providers.
## Issues Resolved
- #497: Core text-to-image support
- #498: Image editing support
- #499: Async API variants
- #500: Additional cloud provider support
- #501: Cost tracking and metrics integration
### New Files (5)
- [ImagePricingRegistry.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/main/scala/org/llm4s/imagegeneration/ImagePricingRegistry.scala:0:0-0:0) - Pricing data for all supported image providers
- `VertexAIClient.scala` - Google Vertex AI Imagen support
- `BedrockClient.scala` - AWS Bedrock Titan Image Generator support
- [StabilityAIClient.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/main/scala/org/llm4s/imagegeneration/provider/StabilityAIClient.scala:0:0-0:0) - Stability AI Direct API support
- [FalAIClient.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/main/scala/org/llm4s/imagegeneration/provider/FalAIClient.scala:0:0-0:0) - Fal AI (Flux/SDXL) support
### Modified Files (12)
- [ImageGeneration.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/main/scala/org/llm4s/imagegeneration/ImageGeneration.scala:0:0-0:0) - Added provider enums, configs, and factory methods
- `StableDiffusionClient.scala`, `HuggingFaceClient.scala`, `OpenAIImageClient.scala` - Enhanced with async support
- [TraceEvent.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/main/scala/org/llm4s/trace/TraceEvent.scala:0:0-0:0) - Added `ImageGenerationCostRecorded` event
- [MetricsCollector.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/main/scala/org/llm4s/metrics/MetricsCollector.scala:0:0-0:0), [PrometheusMetrics.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/main/scala/org/llm4s/metrics/PrometheusMetrics.scala:0:0-0:0) - Added `recordImageGeneration()` method
- [ConsoleTracing.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/main/scala/org/llm4s/trace/ConsoleTracing.scala:0:0-0:0), [LangfuseTracing.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/main/scala/org/llm4s/trace/LangfuseTracing.scala:0:0-0:0), [OpenTelemetryTracing.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/trace-opentelemetry/src/main/scala/org/llm4s/trace/OpenTelemetryTracing.scala:0:0-0:0) - Handle new trace events
- [MockMetricsCollector.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/test/scala/org/llm4s/metrics/MockMetricsCollector.scala:0:0-0:0), [ImageGenerationTest.scala](cci:7://file:///c:/Users/singh/Documents/GitHub/llm4ss/modules/core/src/test/scala/org/llm4s/imagegeneration/ImageGenerationTest.scala:0:0-0:0) - Updated tests
## Features
- **6 Providers**: OpenAI, StableDiffusion, HuggingFace, VertexAI, Bedrock, StabilityAI, FalAI
- **Async APIs**: `generateImageAsync()`, `generateImagesAsync()`, `editImageAsync()`
- **Cost Tracking**: `ImagePricingRegistry` with per-image pricing for all providers
- **Metrics**: Prometheus-compatible metrics for image generation operations
## Testing
- **37 tests passing**
- Unit tests for all configs, factory methods, and pricing lookups
- 
Screenshot 2026-02-09 174618

### CODE REVIEW

**Comment by mandeepsingh2007** [comment]
@rorygraves @vim89  sir i implemented all changes you told please review it:
#549 
#550 
#551 
#552 
#554 
#555

File: modules/core/src/main/scala/org/llm4s/imagegeneration/provider/BedrockClient.scala [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+  private def makeHttpRequest(payload: ujson.Value): Either[ImageGenerationError, requests.Response] = {
+    val payloadBytes = write(payload).getBytes("UTF-8")
+
+
+    val headers = (config.accessKeyId, config.secretAccessKey) match {
+      case (Some(accessKey), Some(secretKey)) =>
+        buildSignedHeaders(payloadBytes, accessKey, secretKey)
+      case _ =>
+        // Fallback to environment credentials
+        val envAccessKey = sys.env.getOrElse("AWS_ACCESS_KEY_ID", "")
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/imagegeneration/provider/BedrockClient.scala:167]: Using sys.env in modules/core main code violates the repository configuration boundary and scalafix policy. Credentials should come only from typed config injected at the app edge. Principle: keep configuration access out of core library code.

File: modules/core/src/main/scala/org/llm4s/imagegeneration/provider/BedrockClient.scala [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        url = apiUrl,
+        data = payloadBytes,
+        headers = headers,
+        readTimeout = config.timeout,
+        connectTimeout = 10000
+      )
+    }.toEither.left.map(e => UnknownError(e))
+  }
+
+  private def buildSignedHeaders(payload: Array[Byte], accessKey: String, secretKey: String): Map[String, String] = {
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/imagegeneration/provider/BedrockClient.scala:189]: This is a hand-rolled SigV4 implementation inside core provider code without focused unit coverage for signing edge cases. Please use AWS SDK signer or a tested signing library and add tests before merge. Principle: avoid custom security critical protocol implementations when standard libraries exist.

File: modules/core/src/main/scala/org/llm4s/imagegeneration/provider/OpenAIImageClient.scala [hunk 3]
```diff
@@ -79,6 +81,108 @@ class OpenAIImageClient(config: OpenAIConfig) extends ImageGenerationClient {
...
+  }
+
+  /**
+   * Generate an image asynchronously
+   */
+  override def generateImageAsync(
+      prompt: String,
+      options: ImageGenerationOptions = ImageGenerationOptions()
+  )(implicit ec: ExecutionContext): Future[Either[ImageGenerationError, GeneratedImage]] =
+    Future {
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/imagegeneration/provider/OpenAIImageClient.scala:157]: This is Future wrapping a blocking HTTP path, so it is concurrency wrapping and not true async IO. Either switch to real async HTTP or remove these async methods from this PR scope. Principle: APIs must match actual runtime behaviour and performance characteristics.

File: modules/core/src/main/scala/org/llm4s/imagegeneration/ImageGeneration.scala [hunk 4]
```diff
@@ -186,13 +218,97 @@ case class OpenAIConfig(
   /** OpenAI API key */
   apiKey: String,
   /** Model to use (dall-e-2 or dall-e-3) */
-  model: String = "dall-e-2",
+  model: String = "gpt-image-1",
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/imagegeneration/ImageGeneration.scala:221]: This changes the default model and silently changes behaviour for existing users of OpenAIConfig and convenience methods. Please keep dall-e-2 as default in this PR and move GPT default switch behind a planned migration path. Principle: preserve backward compatibility for public API defaults.

File: modules/core/src/main/scala/org/llm4s/imagegeneration/ImageGeneration.scala [hunk 5]
```diff
@@ -210,6 +326,35 @@ trait ImageGenerationClient {
     options: ImageGenerationOptions = ImageGenerationOptions()
   ): Either[ImageGenerationError, Seq[GeneratedImage]]
 
+  /** Edit an existing image based on a prompt and optional mask */
+  def editImage(
```
Comment (vim89) [inline modules/core/src/main/scala/org/llm4s/imagegeneration/ImageGeneration.scala:330]: These new trait methods are abstract, so external implementations of ImageGenerationClient will break at compile time. Please add safe default implementations in the trait for compatibility. Principle: extend interfaces without forcing immediate downstream breakage.

**Comment by mandeepsingh2007** [comment]
@vim89 sir please review small pr i am not working in this pr @rorygraves sir told me to open the small pr 
#549 
#550 
#551 
#552 
#554
#555

**Comment by rorygraves** [comment]
Closing as split into smaller PRs
